nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ListFile.java

```diff
@@ -0,0 +1,378 @@
+                } catch (Exception ignore) {
+                } // allow other attributes if these fail
+            }
+            if (store.supportsFileAttributeView("owner")) {
+                try {
+                    FileOwnerAttributeView view = Files.getFileAttributeView(filePath, FileOwnerAttributeView.class);
+                    attributes.put(FILE_OWNER_ATTRIBUTE, view.getOwner().getName());
+                } catch (Exception ignore) {
+                } // allow other attributes if these fail
+            }
+            if (store.supportsFileAttributeView("posix")) {
+                try {
+                    PosixFileAttributeView view = Files.getFileAttributeView(filePath, PosixFileAttributeView.class);
+                    attributes.put(FILE_PERMISSIONS_ATTRIBUTE, PosixFilePermissions.toString(view.readAttributes().permissions()));
+                    attributes.put(FILE_GROUP_ATTRIBUTE, view.readAttributes().group().getName());
+                } catch (Exception ignore) {
+                } // allow other attributes if these fail
+            }
+        } catch (IOException ioe) {
+            // well then this FlowFile gets none of these attributes
+        }
+
+        return attributes;
+    }
+
+    @Override
+    protected String getPath(final ProcessContext context) {
+        return context.getProperty(DIRECTORY).evaluateAttributeExpressions().getValue();
+    }
+
+    @Override
+    protected List<FileInfo> performListing(final ProcessContext context, final Long minTimestamp) throws IOException {
+        final File path = new File(getPath(context));
+        final Boolean recurse = context.getProperty(RECURSE).asBoolean();
+        return scanDirectory(path, fileFilterRef.get(), recurse, minTimestamp);
+    }
+
+    @Override
+    protected boolean isListingResetNecessary(final PropertyDescriptor property) {
+        return DIRECTORY.equals(property)
+                || RECURSE.equals(property)
+                || FILE_FILTER.equals(property)
+                || PATH_FILTER.equals(property)
+                || MIN_AGE.equals(property)
+                || MAX_AGE.equals(property)
+                || MIN_SIZE.equals(property)
+                || MAX_SIZE.equals(property)
+                || IGNORE_HIDDEN_FILES.equals(property);
+    }
+
+    private List<FileInfo> scanDirectory(final File path, final FileFilter filter, final Boolean recurse,
+                                         final Long minTimestamp) throws IOException {
+        final List<FileInfo> listing = new ArrayList<>();
+        File[] files = path.listFiles();
+        if (files != null) {
+            for (File file : files) {
+                if (file.isDirectory()) {
+                    if (recurse) {
+                        listing.addAll(scanDirectory(file, filter, true, minTimestamp));
+                    }
+                } else {
+                    if (filter.accept(file)) {
```

Comment:
So it appears that filter is only applied on the 'file' not 'directory'. Is this intentional?